crates/sui-indexer/src/handlers/checkpoint_handler.rs

```diff
@@ -572,8 +544,137 @@ where
         })
     }
 
+    fn index_packages(
+        transactions: &[CheckpointTransactionBlockResponse],
+        changed_objects: &[(ObjectStatus, SuiObjectData)],
+    ) -> Result<Vec<Package>, IndexerError> {
+        let object_map = changed_objects
+            .iter()
+            .filter_map(|(_, o)| {
+                if let SuiRawData::Package(p) = &o
+                    .bcs
+                    .as_ref()
+                    .expect("Expect the content field to be non-empty from data fetching")
+                {
+                    Some((o.object_id, p))
+                } else {
+                    None
+                }
+            })
+            .collect::<BTreeMap<_, _>>();
+
+        transactions
+            .iter()
+            .flat_map(|tx| {
+                tx.effects.created().iter().map(|oref| {
+                    object_map
+                        .get(&oref.reference.object_id)
+                        .map(|o| Package::try_from(*tx.transaction.data.sender(), o))
+                })
+            })
+            .flatten()
+            .collect()
+    }
+}
+
+struct CheckpointProcessor<S>
+where
+    S: IndexerStore + Clone + Sync + Send + 'static,
+{
+    state: S,
+    metrics: IndexerMetrics,
+    epoch_sender: Arc<Mutex<Sender<TemporaryEpochStore>>>,
+    checkpoint_sender: Arc<Mutex<Sender<TemporaryCheckpointStore>>>,
+    rx: Receiver<CheckpointData>,
+}
+
+impl<S> CheckpointProcessor<S>
+where
+    S: IndexerStore + Clone + Sync + Send + 'static,
+{
+    async fn run(&mut self) -> Result<(), IndexerError> {
+        loop {
+            let checkpoint_data = self
+                .rx
+                .blocking_recv()
+                .expect("Sender of Checkpoint Processor's rx should not be closed.");
+
+            // Index checkpoint data
+            let index_timer = self.metrics.checkpoint_index_latency.start_timer();
+
+            let (checkpoint, epoch) =
+                Self::index_checkpoint_and_epoch(&self.state, &checkpoint_data)
+                    .await
+                    .tap_err(|e| {
+                        error!(
+                            "Failed to index checkpoints {:?} with error: {}",
+                            checkpoint_data,
+                            e.to_string()
+                        );
+                    })?;
+            index_timer.stop_and_record();
+
+            // commit first epoch immediately, send other epochs to channel to be committed later.
+            if let Some(epoch) = epoch {
```

Comment:
not this PR, I also wanted to move all epoch handling to a separate async processor, b/c it's less frequent and not on the critical path, which will also make codes here cleaner.

Reply:
agreed